### ISSUE TITLE
Quaternion/Matrix issue after upgrading from 2.0 to 3.0

### ISSUE THREAD

**Issue by RUSshy** [opentk/opentk#774]
Hi,

I tried to upgrade a project from opentk 2.0 to 3.0

But i run into issue

I don't know what exacly changed/broke but it seems to be related to Matrix/Quarternions

Here is the project in question: https://github.com/RUSshy/research-skeletal-animation

Expected result:

- Model should render correctly

Result i got:

- Model isn't rendered correctly

[screenshot](https://i.imgur.com/bb3OR3M.png)

**Comment by RUSshy** [comment]
I tried with OpenTK 3.0pre and issue is not present

**Comment by Nihlus** [comment]
We need more information than this.

**Comment by RUSshy** [comment]
Well, the information is here, upgrade from 2.0 to 3.0

There is no mention of breaking change in the readme or https://github.com/opentk/opentk/releases, so it should work without change

**Comment by Nihlus** [comment]
That doesn't help us track down the issue - we're a very small team of volunteers, and we unfortunately don't have time to go through your source code to find the problem. There are a wide number of other applications using the latest version of OpenTK without issue, so it's most likely an edge case that you've hit.

In order to help you, we need more information about what you're doing, how you're doing it, and where the problem is occuring.

**Comment by RUSshy** [comment]
Oh yes i understand, i will try to track down what's exactly not working correctly then

**Comment by RUSshy** [comment]
What commit revision correspond to the release of 3.0pre ?

This will help me to know where to look since that's the latest build that is working as intended for me

**Comment by RUSshy** [comment]
I might be wrong but i think the problem is in ``Matrix.CreateFromQuaternion``

Quaternion seems to have changed recently: https://github.com/opentk/opentk/commit/fca36b19b06c04e7ba283c2678d2148a37333cbc#diff-2366f01bb4a08f8cd331d42d2cbf9eba


But the code in the Matrix class for quaternion doesn't seems to take into account the new change

**Comment by RUSshy** [comment]
Oh my bad i mean in ``Matrix4.ExtractRotation``

**Comment by ThomasD13** [comment]
I changed the API. See https://github.com/opentk/opentk/pull/714 for more information.

**Comment by RUSshy** [comment]
Thanks for the information!

I tried do some some changes where it"s dealing with Quaternion but i can't seems to fix that issue, there must be something else..

**Comment by TominoCZ** [comment]
I'm also having issues with Quaternions...
I am using this code to get the camera look direction in my game:
`Vector3.Transform(Vector3.UnitX, new Quaternion(-_pitch, -_yaw + MathHelper.PiOver2, 0));`

That worked perfectly until the Quaternion PR..

This is what it looked like before:
![2018-7-2_22 57 52](https://user-images.githubusercontent.com/24359011/42190298-9c5229fa-7e5c-11e8-963a-508c1ffcd76f.png)
![2018-7-2_22 57 49](https://user-images.githubusercontent.com/24359011/42190299-9c707630-7e5c-11e8-94b8-4e10f965e60a.png)

This is what it looks like now:
![2018-7-2_21 44 39](https://user-images.githubusercontent.com/24359011/42190314-b090e046-7e5c-11e8-9687-ada06bb47166.png)
![2018-7-2_21 54 55](https://user-images.githubusercontent.com/24359011/42190315-b0ac4840-7e5c-11e8-848f-bb32128f6430.png)
![youspinmeround](https://user-images.githubusercontent.com/24359011/42190551-06a3b016-7e5e-11e8-8bde-38a68e252d43.gif)

**Comment by ThomasD13** [comment]
Thanks for providing screenshots TominoCZ. I can hardly tell whats wrong in your screenshots. Neither I am a minecraft player ;)

Could you please provide screenshots with and without OpenTK 3.0, at exact the same position in the game, and viewinig direction of the player? Maybe right after the initial loading?

Please read the last comment by me in https://github.com/opentk/opentk/pull/714 (6. Januar)
There is a summary which API I changed and what's the before/after behaviour.
Double check your code against this. If you have still an issue give me a boiled down example and I will investigate this.

**Comment by TominoCZ** [comment]
I use Quaternions for simple rotation of vertexes and also for the camera look direction vector(so that I have the item in front of the camera)

I don't know what's wrong either.. 

I thought I could just use Quaternions to simply rotate a vector but now it doesn't seem right...
I also thought that if this doesn't work I could use `Quaternion.FromEulerAngles()` so I could use pitch/yaw/roll but it calls the same constructor....

I made a playlist of some test videos, I hope it will be useful: 
https://www.youtube.com/playlist?list=PLVg4mPEYqTtc1Ndl_zUiPdf78e4zg_Xrj

From what I've seen, the Z value of the direction vector can get down to `-2.0` in OpenTK 3.0, which isn't right... I'm rotating a unit vector.. it shouldn't return anything longer than 1..

**Comment by ThomasD13** [comment]
@TominoCZ:
From your video https://www.youtube.com/watch?v=Bo0XzOpUK0g&list=PLVg4mPEYqTtc1Ndl_zUiPdf78e4zg_Xrj&index=5&t=0s

at 15 seconds, I can see that Pitch 0 and Yaw 274 result in -1.00 0.00 and -2.00. 
What is this vector? (-1.00, 0.00, -2.00). And how is this calculated? Can you provide me that piece of Code?

**Comment by TominoCZ** [comment]
`Vector3.Transform(Vector3.UnitX, new Quaternion(-_pitch, -_yaw + MathHelper.PiOver2, 0));`

I mentioned that in the first comment I added btw.

**Comment by ThomasD13** [comment]
Thats interesting. Indeed: Rotating a vector should not change it's length. This could be a missed leftover from the old code, and a new potential bug. I will investigate this. 
Should be easy reproducable. Thanks for your information @TominoCZ .

**Comment by TominoCZ** [comment]
@ThomasD13 I tried using the Quaternion code from 2.0 for rotation to see if it works but it actually didn't.

I have made this function which is basically Quaternion code from 2.0 for rotating a vector:
```
        public static Vector3 Rotate(Vector3 vec, float pitch, float yaw, float roll)
        {
            yaw *= 0.5f;
            pitch *= 0.5f;
            roll *= 0.5f;
            float num1 = (float) Math.Cos((double) yaw);
            float num2 = (float) Math.Cos((double) pitch);
            float num3 = (float) Math.Cos((double) roll);
            float num4 = (float) Math.Sin((double) yaw);
            float num5 = (float) Math.Sin((double) pitch);
            float num6 = (float) Math.Sin((double) roll);

            Vector3 xyz;
            var w = (float) ((double) num1 * (double) num2 * (double) num3 - (double) num4 * (double) num5 * (double) num6);
            xyz.X = (float) ((double) num4 * (double) num5 * (double) num3 + (double) num1 * (double) num2 * (double) num6);
            xyz.Y = (float) ((double) num4 * (double) num2 * (double) num3 + (double) num1 * (double) num5 * (double) num6);
            xyz.Z = (float) ((double) num1 * (double) num5 * (double) num3 - (double) num4 * (double) num2 * (double) num6);

            Vector3 result;
            Vector3 result1;
            Vector3.Cross(ref xyz, ref vec, out result1);
            Vector3 result2;
            Vector3.Multiply(ref vec, w, out result2);
            Vector3.Add(ref result1, ref result2, out result1);
            Vector3.Cross(ref xyz, ref result1, out result1);
            Vector3.Multiply(ref result1, 2f, out result1);
            Vector3.Add(ref vec, ref result1, out result);

            return result;
        }
```

While still using 3.0 with this little function with code from 2.0, I got the same exact results.
When I switch back to 2.0 while still using this same function, it works as it should.

I mean.. honestly, the only thing that could've changed between different versions of OpenTK in this piece of code are the **Vector3** functions. Maybe the problem is there.

**Comment by ThomasD13** [comment]
Thats a good hint. 
Just to be sure: So the first part of your Rotate-Method is from the 2.0 Quaternion class? And the second part is from the Vector3 Transform Method, which is identical between 2.0 and 3.0 (I havn't checked this yet). 
So with your method, you excluded my Quaternion fix. Moreover you excluded the Quaternion-functionality depending of any particular OpenTK verison.
In my opinion, there must be a change in one Vector3-Method which is responsible for your wrong results. I will check this at next weekend.

**Comment by TominoCZ** [comment]
Exactly :)

**Comment by TominoCZ** [comment]
@ThomasD13 @RUSshy fixed: #777, #776

**Comment by RUSshy** [comment]
wow thanks a lot @TominoCZ !!!!

When will nuget package will be updated ?

**Comment by TominoCZ** [comment]
@RUSshy I'm not from the official dev team so idk :'D

I made this function as a temporary fix before release 4.0:
```
public static Vector3 Rotate(Vector3 vec, float pitch, float yaw, float roll)
        {
            //Quaternion constructor
            yaw *= 0.5f;
            pitch *= 0.5f;
            roll *= 0.5f;
            float num1 = (float) Math.Cos((double) yaw);
            float num2 = (float) Math.Cos((double) pitch);
            float num3 = (float) Math.Cos((double) roll);
            float num4 = (float) Math.Sin((double) yaw);
            float num5 = (float) Math.Sin((double) pitch);
            float num6 = (float) Math.Sin((double) roll);

            Vector3 xyz;
            var w = (float) ((double) num1 * (double) num2 * (double) num3 - (double) num4 * (double) num5 * (double) num6);
            xyz.X = (float) ((double) num4 * (double) num5 * (double) num3 + (double) num1 * (double) num2 * (double) num6);
            xyz.Y = (float) ((double) num4 * (double) num2 * (double) num3 + (double) num1 * (double) num5 * (double) num6);
            xyz.Z = (float) ((double) num1 * (double) num5 * (double) num3 - (double) num4 * (double) num2 * (double) num6);
            
            //fixed Vector3.Transform(Quaternion)
            Vector3 result, temp, temp2;
            Vector3.Cross(ref xyz, ref vec, out temp);
            Vector3.Multiply(ref vec, w, out temp2);
            Vector3.Add(ref temp, ref temp2, out temp);
            Vector3.Cross(ref xyz, ref temp, out temp2);
            Vector3.Multiply(ref temp2, 2f, out temp2);
            Vector3.Add(ref vec, ref temp2, out result);

            return result;
        }
```

**Comment by Nihlus** [comment]
Wonderful work, all of you. I've merged the pull request, and I'm pushing out a hotfix release today. NuGet should update with the new version with the fix within a few hours.
